### PR TITLE
[message_filters] Fast approximate time synchronization in message_filters (pure python)

### LINKED ISSUE #800: [message_filters] Python interface of approximate synchronizer hangs up with large queue size
## Issue

Topic's fps goes down. When I use the message_filter's approximate synchronizer with large queue size (100) and processing in the callback which takes time (1 second).

```
% roslaunch jsk_20160505_test_message_filters string.launch

% rostopic hz /join_strings/output
subscribed to [/join_strings/output]
average rate: 0.997
        min: 1.003s max: 1.003s std dev: 0.00000s window: 2
average rate: 0.996
        min: 1.003s max: 1.004s std dev: 0.00019s window: 3
average rate: 0.975
        min: 1.003s max: 1.070s std dev: 0.03121s window: 4
average rate: 0.962
        min: 1.003s max: 1.079s std dev: 0.03554s window: 5
average rate: 0.949
        min: 1.003s max: 1.112s std dev: 0.04305s window: 6
average rate: 0.925
        min: 1.003s max: 1.219s std dev: 0.07304s window: 7
average rate: 0.915
        min: 1.003s max: 1.219s std dev: 0.07346s window: 8
no new messages
average rate: 0.887
        min: 1.003s max: 1.373s std dev: 0.11528s window: 9
average rate: 0.871
        min: 1.003s max: 1.373s std dev: 0.12357s window: 10
no new messages
average rate: 0.786
        min: 1.003s max: 2.381s std dev: 0.38777s window: 11
no new messages
no new messages
no new messages
no new messages
average rate: 0.621
        min: 1.003s max: 5.008s std dev: 1.13588s window: 12
no new messages
no new messages
average rate: 0.579
        min: 1.003s max: 5.008s std dev: 1.15229s window: 13
no new messages
no new messages
no new messages
no new messages
average rate: 0.519
        min: 1.003s max: 5.008s std dev: 1.30438s window: 14
no new messages
no new messages
no new messages
average rate: 0.470
        min: 1.003s max: 5.008s std dev: 1.45689s window: 15
no new messages
```
## Testing Code
- Package: https://github.com/wkentaro/jsk_20160505_test_message_filters
- string.launch: https://github.com/wkentaro/jsk_20160505_test_message_filters/blob/master/launch/string.launch
- publish_string: https://github.com/wkentaro/jsk_20160505_test_message_filters/blob/master/scripts/publish_string
- join_strings: https://github.com/wkentaro/jsk_20160505_test_message_filters/blob/master/scripts/join_strings  (sync 3 topics with 1sec sleep in callback)
## Solution

After some workaround with print debugging, I found this section takes time when the program hangs up.
https://github.com/ros/ros_comm/blob/kinetic-devel/utilities/message_filters/src/message_filters/__init__.py#L224-L231 
So solutions are:
- Refactor code for searching the timestamp to do it with pure python #802 
- Use numpy for searching the timestamp #801

### PR DESCRIPTION
**Pull request by wkentaro**
Closes #800 
## Result

```
% rostopic hz /join_strings/output
subscribed to [/join_strings/output]
average rate: 0.995
        min: 1.005s max: 1.005s std dev: 0.00000s window: 2
average rate: 0.932
        min: 1.005s max: 1.141s std dev: 0.06805s window: 3
average rate: 0.953
        min: 1.002s max: 1.141s std dev: 0.06469s window: 4
average rate: 0.963
        min: 1.002s max: 1.141s std dev: 0.05930s window: 5
average rate: 0.950
        min: 1.002s max: 1.141s std dev: 0.06099s window: 6
average rate: 0.936
        min: 1.002s max: 1.144s std dev: 0.06516s window: 7
average rate: 0.944
        min: 1.002s max: 1.144s std dev: 0.06444s window: 8
average rate: 0.948
        min: 1.002s max: 1.144s std dev: 0.06143s window: 9
average rate: 0.953
        min: 1.002s max: 1.144s std dev: 0.05997s window: 10
no new messages
average rate: 0.955
        min: 1.002s max: 1.144s std dev: 0.05706s window: 11
average rate: 0.958
        min: 1.002s max: 1.144s std dev: 0.05597s window: 12
average rate: 0.962
        min: 1.002s max: 1.144s std dev: 0.05481s window: 13
average rate: 0.964
        min: 1.002s max: 1.144s std dev: 0.05361s window: 14
average rate: 0.967
        min: 1.002s max: 1.144s std dev: 0.05243s window: 15
average rate: 0.969
        min: 1.002s max: 1.144s std dev: 0.05129s window: 16
average rate: 0.970
        min: 1.002s max: 1.144s std dev: 0.05019s window: 17
average rate: 0.972
        min: 1.002s max: 1.144s std dev: 0.04914s window: 18
average rate: 0.973
        min: 1.002s max: 1.144s std dev: 0.04802s window: 19
average rate: 0.974
        min: 1.002s max: 1.144s std dev: 0.04697s window: 20
```

### CODE REVIEW

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder retest this please

**Comment by wkentaro** [comment]
I removed the debugging lines, please review and give me feedback when you have time.

File: utilities/message_filters/src/message_filters/__init__.py [hunk 1]
```diff
@@ -184,9 +185,11 @@ def __init__(self, fs, queue_size):
 
     def connectInput(self, fs):
         self.queues = [{} for f in fs]
-        self.input_connections = [f.registerCallback(self.add, q) for (f, q) in zip(fs, self.queues)]
+        self.input_connections = [
+            f.registerCallback(self.add, q, i_q)
+            for i_q, (f, q) in enumerate(zip(fs, self.queues))]
 
-    def add(self, msg, my_queue):
+    def add(self, msg, my_queue, my_queue_index):
```
Comment (dirk-thomas) [inline utilities/message_filters/src/message_filters/__init__.py]: This changes existing public API and would therefore break existing code. I don't think this is feasible for an already released ROS distro.

**Comment by wkentaro** [comment]
Updated.

**Comment by wkentaro** [comment]
Could you please give me some comments?

**Comment by dirk-thomas** [comment]
I am sorry for the late response but I currently have to focus on a different project and don't have much time to review ROS tickets.

The patch looks good to me. Thank you for iterating on it.

**Comment by wkentaro** [comment]
I see. Thank you.

**Pull request merged**